pkg/terraform/workspace_test.go

```diff
@@ -0,0 +1,454 @@
+	changeSummaryUpdate   = `{"@level":"info","@message":"Plan: 0 to add, 1 to change, 0 to destroy.","@module":"terraform.ui","@timestamp":"0000-00-00T00:00:00.000000+03:00","changes":{"add":0,"change":1,"remove":0,"operation":"plan"},"type":"change_summary"}`
+	changeSummaryNoAction = `{"@level":"info","@message":"Plan: 0 to add, 0 to change, 0 to destroy.","@module":"terraform.ui","@timestamp":"0000-00-00T00:00:00.000000+03:00","changes":{"add":0,"change":0,"remove":0,"operation":"plan"},"type":"change_summary"}`
+
+	state = &json.StateV4{
+		Version:          uint64(version),
+		TerraformVersion: terraformVersion,
+		Serial:           uint64(serial),
+		Lineage:          lineage,
+		RootOutputs:      map[string]json.OutputStateV4{},
+		Resources:        []json.ResourceStateV4{},
+	}
+)
+
+func newFailureWorkspace() *Workspace {
+	return NewWorkspace(directory, WithExec(&testingexec.FakeExec{
+		CommandScript: []testingexec.FakeCommandAction{
+			func(_ string, _ ...string) k8sExec.Cmd {
+				b := bytes.Buffer{}
+				b.WriteString(errBoom.Error())
+
+				return &testingexec.FakeCmd{
+					CombinedOutputScript: []testingexec.FakeAction{
+						func() ([]byte, []byte, error) {
+							return b.Bytes(), nil, errBoom
+						},
+					},
+				}
+			},
+		},
+	},
+	), WithLogger(newLogger()))
+}
+
+func newWorkspaceWithChangeSummary(changeSummary string) *Workspace {
+	return NewWorkspace(
+		directory, WithExec(&testingexec.FakeExec{
+			CommandScript: []testingexec.FakeCommandAction{
+				func(_ string, _ ...string) k8sExec.Cmd {
+					b := bytes.Buffer{}
+					b.WriteString(changeSummary)
+
+					return &testingexec.FakeCmd{
+						CombinedOutputScript: []testingexec.FakeAction{
+							func() ([]byte, []byte, error) {
+								return b.Bytes(), nil, nil
+							},
+						},
+					}
+				},
+			},
+		},
+		), WithLogger(newLogger()))
+}
+
+func newLogger() logging.Logger {
+	zl := zap.New(zap.UseDevMode(true))
+	return logging.NewLogrLogger(zl.WithName("provider-jet-test"))
+}
+
+func timeNow() *time.Time {
+	t := time.Unix(0, 0)
+	return &t
+}
+
+func TestWorkspaceApply(t *testing.T) {
+	type args struct {
+		w *Workspace
+	}
+	type want struct {
+		r   ApplyResult
+		err error
+	}
+
+	cases := map[string]struct {
+		args
+		want
+	}{
+		"Running": {
+			args: args{
+				w: NewWorkspace(directory, WithLastOperation(&Operation{Type: testType, startTime: timeNow(), endTime: nil})),
+			},
+			want: want{
+				err: errors.Errorf("%s operation that started at %s is still running", testType, timeNow().String()),
+			},
+		},
+		"Success": {
+			args: args{
+				w: NewWorkspace(
+					directory, WithExec(&testingexec.FakeExec{DisableScripts: true}), WithLogger(newLogger())),
+			},
+			want: want{
+				r: ApplyResult{
+					State: state,
+				},
+			},
+		},
+		"Failure": {
+			args: args{
+				w: newFailureWorkspace(),
```

Comment:
I wonder if we can come up with a function that is capable of being called for all cases in a table so that it's more consistent and easy to read. It'd be awesome if we'd be able to use existing `NewWorkspace` with function options though that could be a bit tricky with afero filesystem where you'd need a function context to fill the files in the memory.

Reply:
Another option could be to accept more granular things, like `executor`, filesystem and possible last operation and then use them to initialize `NewWorkspace` in `t.Run`

Reply:
I tried to create `Workspace` objects by using the `NewWorkspace` and its `options`. For creating `FakeExec`, I have added a utility function and I passed this to the `WithExecutor`. For filling the `afero` file, in `t.Run`, `WriteFile` was used.